plenoptic/synthesize/geodesic.py

```diff
@@ -1,155 +1,304 @@
+        # signal_grad = x.grad
+        # x.grad = signal_grad - (
+        #           signal_grad @ repres_grad
+        #                         ) / torch.norm(repres_grad) * repres_grad
 
-        z = x.permute(1, 0, 2, 3).unsqueeze(1)
-        loss = 0
-        for s in range(self.n_scales):
-            loss = loss + torch.sum(self.diff(z) ** 2)
-            z = self.blur(z)
+        self.optimizer.step()
+        grad_norm = torch.norm(self.x.grad.data)
+        pbar.set_postfix(OrderedDict([('loss', f'{loss.item():.4e}'),
+                         ('gradient norm', f'{grad_norm:.4e}'),
+                         ('lr', self.optimizer.param_groups[0]['lr'])]))
+        if grad_norm.item() != grad_norm.item():
+            raise Exception('found a NaN in the gradients during optimization')
 
         return loss
 
-    def _optimizer_step(self, i, max_iter):
+    def synthesize(self, max_iter=1000, learning_rate=.001, optimizer='adam',
+                   lmbda=.1, objective='multiscale', seed=0):
+        """ synthesize a geodesic
 
-        self.optimizer.zero_grad()
-        output = self.analyze(self.x)
-        loss = self.objective_function(output)
-        loss.backward()
-        g = self.x.grad.data
-        if math.isnan(loss.item()):
-            raise Exception('found a NaN during optimization')
-        self.optimizer.step()
+        Parameters
+        ----------
+        max_iter: int, optional
+            maximum number of steps taken by the optimization a
 
-        self.t0 = self.t1
-        self.t1 = time.time()
-        pretty_print(i, max_iter, (self.t1 - self.t0), loss.item(), g.norm().item())
+        learning_rate: float, optional
+            controls the step sizes of the search algorithm
 
-        return loss
+        optimizer: str ('adam' or 'sgd') or torch.optim.Optimizer, optional
+            algorithm that will perform the search
 
-    def synthesize(self, max_iter=20, learning_rate=.01, objective='multiscale', seed=0):
+        lmbda: float, optional
+            strength of the regularizer that enforces the image range in [0, 1]
 
-        torch.manual_seed(seed)
-        self.optimizer = optim.Adam([self.x], lr=learning_rate, amsgrad=True)
+        objective: str, optional
+            'default'
+            'multiscale'
 
-        self.loss = []
-        self.t1 = time.time()
+        seed: int
+            set the random number generator
+        """
+        self.lmbda = lmbda
 
-        for i in range(max_iter):
-            loss = self._optimizer_step(i, max_iter)
+        torch.manual_seed(seed)
+        if optimizer == 'adam':
+            self.optimizer = optim.Adam([self.x],
+                                        lr=learning_rate, amsgrad=True)
+        elif optimizer == 'sgd':
+            self.optimizer = optim.SGD([self.x],
+                                       lr=learning_rate, momentum=0.9)
+        elif isinstance(optimizer, optim.Optimizer):
+            self.optimizer = optimizer
```

Comment:
I'm guessing you're of the opinion that if the user wants to change the kwargs for SGD or Adam, they should pass their own optimizer. That's fine but we should: 
- make explicit what they are in docstring / documentation somewhere
- be consistent. I currently set `nesterov=True` and `momentum=.8` by default for SGD (in `Synthesis`), so why'd you decide on these values?

Reply:
and if we're not consistent, we should have clear reasoning as to why